oxia-operator/controllers/oxiacluster_coordinator.go

```diff
@@ -0,0 +1,218 @@
+// limitations under the License.
+
+package controllers
+
+import (
+	"context"
+	monitoringV1 "github.com/prometheus-operator/prometheus-operator/pkg/apis/monitoring/v1"
+	oxiav1alpha1 "github.com/streamnative/oxia/api/v1alpha1"
+	appsV1 "k8s.io/api/apps/v1"
+	coreV1 "k8s.io/api/core/v1"
+	rbacV1 "k8s.io/api/rbac/v1"
+	apiErrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+)
+
+type ServerAddress struct {
+	// Public is the endpoint that is advertised to clients
+	Public string `json:"public" yaml:"public"`
+	// Internal is the endpoint for server->server RPCs
+	Internal string `json:"internal" yaml:"internal"`
+}
+
+type ClusterConfig struct {
+	Namespaces []NamespaceConfig `json:"namespaces" yaml:"namespaces"`
+	Servers    []ServerAddress   `json:"servers" yaml:"servers"`
+}
+
+type NamespaceConfig struct {
+	Name              string `json:"name" yaml:"name"`
+	InitialShardCount uint32 `json:"initialShardCount" yaml:"initialShardCount"`
+	ReplicationFactor uint32 `json:"replicationFactor" yaml:"replicationFactor"`
+}
+
+var CoordinatorPorts = []NamedPort{InternalPort, MetricsPort}
+
+func (r *OxiaClusterReconciler) reconcileCoordinator(ctx context.Context, oxia *oxiav1alpha1.OxiaCluster) (ctrl.Result, error) {
+	// Reconcile service account
+	_serviceAccount := &coreV1.ServiceAccount{}
+	nsName := types.NamespacedName{Name: resourceName(Coordinator, oxia.Name), Namespace: oxia.Namespace}
+	err := r.Client.Get(ctx, nsName, _serviceAccount)
+	if err != nil {
+		if !apiErrors.IsNotFound(err) {
+			r.Log.Error(err, "Failed to get service account.")
+			return ctrl.Result{}, err
+		}
+		r.Log.Info("Creating a new service account.", "NamespaceName", nsName)
+		err := r.Client.Create(ctx, serviceAccount(Coordinator, oxia))
+		if err != nil {
+			r.Log.Error(err, "Failed to create service account :", "NamespaceName", nsName)
+			return ctrl.Result{}, err
+		}
+		r.Log.Info("Created service account: ", "NamespaceName", nsName)
+		return ctrl.Result{Requeue: true}, nil
+	}
+
+	if oxia.DeletionTimestamp != nil {
+		r.Log.Info("Service account deletion is unsupported yet!")
+	}
+
+	// Reconcile role
+	_role := &rbacV1.Role{}
+	nsName = types.NamespacedName{Name: resourceName(Coordinator, oxia.Name), Namespace: oxia.Namespace}
+	err = r.Client.Get(ctx, nsName, _role)
+	if err != nil {
+		if !apiErrors.IsNotFound(err) {
+			r.Log.Error(err, "Failed to get role: ", "NamespaceName", nsName)
+			return ctrl.Result{}, err
+		}
+		r.Log.Info("Creating a new role :", "NamespaceName", nsName)
+		err := r.Client.Create(ctx, role(Coordinator, oxia))
+		if err != nil {
+			r.Log.Error(err, "Failed to create role:", "NamespaceName", nsName)
+			return ctrl.Result{}, err
+		}
+		r.Log.Info("Created role: ", "NamespaceName", nsName)
+		return ctrl.Result{Requeue: true}, nil
+	}
+
+	if oxia.DeletionTimestamp != nil {
```

Comment:
Does it make sense that if resources are configured with ownerReference, these resources will be automatically cleaned up